Stored filter expressions

signedav opened the pull request:
In the attribute table we can store expressions for filters - load them, edit them and delete them.

![storedfilterexpressions](https://user-images.githubusercontent.com/28384354/63512424-1aa32d80-c4e4-11e9-96f8-505d9544db8e.gif)
(second save action is "Save expression as...")

Regarding QEP https://github.com/qgis/QGIS-Enhancement-Proposals/issues/152

signedav committed:
stored expression base class with read write methods

signedav committed:
read write storedexpressions

signedav committed:
stored expressions comments

signedav committed:
menu basics including stored filter expression

signedav committed:
improve gui for stored filter expressions

signedav committed:
stored filter expression saving part

signedav committed:
remove expression functionality identified by name

signedav committed:
stored expression icons to 'bookmark' expressions

signedav committed:
use of struct instead of QPair and use Uuid instead of name as identification

signedav committed:
on textchange update the actions

signedav committed:
remove unused functions - but having issue when toggling the storeExpressionButton (fires Actions every time)

signedav committed:
descriptive name for mStoredFilterExpressions to make a difference to future QgsStoredExpression objects

signedav committed:
listening on triggered instead of toggled, because triggered is happening on click and not on setChecked etc.

signedav committed:
functionality to save and update stored expressions

signedav committed:
wait 300 ms before update stored expression button after text changed on filter query line edit
more comments and clean up

signedav committed:
make proper constructor

signedav committed:
no constructor in QgsStoredExpression because it leaded to issues with compiling sip

dialoges for stored filter expressions contains labels and expression line edit

signedav committed:
some tests for QgsStoredExpressionManager

m-kuhn commented on src/app/qgsattributetabledialog.h:
```diff
@@ -230,12 +264,15 @@ class APP_EXPORT QgsAttributeTableDialog : public QDialog, private Ui::QgsAttrib
     QStringList mVisibleFields;
     QgsAttributeEditorContext mEditorContext;
 
+    QTimer *mFilterQueryTimer;
```
This will be leaked, no?

I think you don't need multiple of them in parallel. You just need one which is restarted on key press.

```suggestion
    QTimer mFilterQueryTimer;
```

nyalldawson replied:
Also should be initialized here to nullptr if it IS required

m-kuhn commented on src/core/CMakeLists.txt:
```diff
@@ -1011,6 +1013,7 @@ SET(QGIS_CORE_HDRS
   qgssqlexpressioncompiler.h
   qgssqlstatement.h
   qgsstatisticalsummary.h
+  qgsstoredexpressionmanager.h
```
```suggestion
```
I think adding it only in MOC_HDRS is sufficient

m-kuhn commented on src/core/qgsstoredexpressionmanager.cpp:
```diff
@@ -0,0 +1,180 @@
+{
+  Q_UNUSED( tag );
+
+  /* leads to trouble with sip...
```
Can you elaborate?

m-kuhn commented on src/core/qgsstoredexpressionmanager.cpp:
```diff
@@ -0,0 +1,180 @@
+  {
+    if ( storedExpression.id == id )
+    {
+      mStoredExpressions.removeAt( i );
```
Watch out: you are modifying the container you are looping over!
This only works because you have a break below.
Normally you need to take a (const) copy of the container if modifying the source container in the loop.
So best to leave a short comment why it's allowed to do that here.

m-kuhn commented on src/core/qgsstoredexpressionmanager.cpp:
```diff
@@ -0,0 +1,180 @@
+  Q_UNUSED( tag );
+
+  int i = 0;
+  for ( const QgsStoredExpression &storedExpression : mStoredExpressions )
```
```suggestion
  for ( const QgsStoredExpression &storedExpression : qgis::as_const( mStoredExpressions ) )
```

m-kuhn commented on src/core/qgsstoredexpressionmanager.cpp:
```diff
@@ -0,0 +1,180 @@
+  Q_UNUSED( tag );
+
+  int i = 0;
+  for ( const QgsStoredExpression &storedExpression : mStoredExpressions )
```
```suggestion
  for ( const QgsStoredExpression &storedExpression : qgis::as_const( mStoredExpressions ) )
```

m-kuhn commented on src/core/qgsstoredexpressionmanager.cpp:
```diff
@@ -0,0 +1,180 @@
+  mStoredExpressions.clear();
+}
+
+bool QgsStoredExpressionManager::writeXml( QDomNode &layer_node ) const
```
layerNode

m-kuhn commented on src/core/qgsstoredexpressionmanager.cpp:
```diff
@@ -0,0 +1,180 @@
+{
+  Q_UNUSED( tag );
+
+  for ( const QgsStoredExpression &storedExpression : mStoredExpressions )
```
```suggestion
  for ( const QgsStoredExpression &storedExpression : qgis::as_const( mStoredExpressions ) )
```

m-kuhn commented on src/core/qgsstoredexpressionmanager.cpp:
```diff
@@ -0,0 +1,180 @@
+
+  QDomElement aStoredExpressions = layer_node.ownerDocument().createElement( QStringLiteral( "%1s" ).arg( elementName ) );
+
+  for ( const QgsStoredExpression &storedExpression : mStoredExpressions )
```
...

m-kuhn commented on src/core/qgsstoredexpressionmanager.cpp:
```diff
@@ -0,0 +1,180 @@
+  return QgsStoredExpression();
+}
+
+QgsStoredExpression QgsStoredExpressionManager::findStoredExpressionByExpression( const QString &expression, const QString &tag )
```
```suggestion
QgsStoredExpression QgsStoredExpressionManager::findStoredExpressionByExpression( const QString &expression, const QString &tag ) const
```

This shouldn't modify the internals?

m-kuhn commented on src/core/qgsstoredexpressionmanager.cpp:
```diff
@@ -0,0 +1,180 @@
+  return mStoredExpressions;
+}
+
+QgsStoredExpression QgsStoredExpressionManager::storedExpression( const QUuid &id, const QString &tag )
```
```suggestion
QgsStoredExpression QgsStoredExpressionManager::storedExpression( const QUuid &id, const QString &tag ) const
```

m-kuhn commented on src/core/qgsstoredexpressionmanager.h:
```diff
@@ -0,0 +1,158 @@
+struct CORE_EXPORT QgsStoredExpression
+{
+
+  /* leads to trouble with sip...
```
Now I understand the above comment.
What's the trouble?
Can we worst case avoid it with `#ifndef SIP_RUN`? The constructor would be handy if it is C++ only too.

m-kuhn commented on src/core/qgsstoredexpressionmanager.h:
```diff
@@ -0,0 +1,158 @@
+     * FilterExpression for stored expressions to filter attribute table
+     * DefaultValueExpression for stored expressions used for default values could be an extension
+     */
+    enum Mode
```
(How) does that relate to the tags?

m-kuhn reviewed (commented):
Nice!
Some questions here and there but overall I like it :+1:

nyalldawson commented on src/app/qgsattributetabledialog.cpp:
```diff
@@ -1043,6 +1112,92 @@ void QgsAttributeTableDialog::openConditionalStyles()
+void QgsAttributeTableDialog::saveAsStoredFilterExpression()
+{
+  QgsDialog *dlg = new QgsDialog( this, nullptr, QDialogButtonBox::Save | QDialogButtonBox::Cancel );
+  dlg->setWindowTitle( tr( "Save expression as ..." ) );
```
```suggestion
  dlg->setWindowTitle( tr( "Save Expression As" ) );
```

nyalldawson commented on src/app/qgsattributetabledialog.cpp:
```diff
@@ -1043,6 +1112,92 @@ void QgsAttributeTableDialog::openConditionalStyles()
+  QgsDialog *dlg = new QgsDialog( this, nullptr, QDialogButtonBox::Save | QDialogButtonBox::Cancel );
+  dlg->setWindowTitle( tr( "Save expression as ..." ) );
+  QVBoxLayout *layout = dlg->layout();
+  dlg->resize( 400, dlg->height() );
```
This isn't hidpi friendly -- better to save and restore the geometry

nyalldawson commented on src/app/qgsattributetabledialog.h:
```diff
@@ -175,6 +177,35 @@ class APP_EXPORT QgsAttributeTableDialog : public QDialog, private Ui::QgsAttrib
     void filterQueryAccepted();
+
+    /**
+    * handle the expression (save or delete) when the bookmark button for stored
```
Capitalization, full stops at end of sentences. Also we follow the Qt convention of "HandleS..." / "OpenS...

nyalldawson commented on src/core/qgsstoredexpressionmanager.h:
```diff
@@ -0,0 +1,158 @@
+/***************************************************************************
+                          qgsstoredexpressionmanager.h
+
+ These classes store and control the management and execution of actions
```
Instead of putting the nice details up here, but them down in the actual class dox (also, update them to be relevant ;) )

nyalldawson commented on src/core/qgsstoredexpressionmanager.h:
```diff
@@ -0,0 +1,158 @@
+class QDomNode;
+class QDomDocument;
+
+struct CORE_EXPORT QgsStoredExpression
```
Dox

nyalldawson commented on src/core/qgsstoredexpressionmanager.h:
```diff
@@ -0,0 +1,158 @@
+  {}
+  */
+
+  QUuid id;
```
I think instead of using raw QUUid's here, it would be nicer to use a QString in the public API. You can use QUuid in the background to generate initial values for the strings, but only expose them publicly as strings.

nyalldawson commented on src/core/qgsstoredexpressionmanager.h:
```diff
@@ -0,0 +1,158 @@
+    *  \param tag               some content, maybe scope where to be shown
+    *  \returns generated id as QUuid
+    */
+    QUuid addStoredExpression( const QString &name, const QString &expression, const QString &tag = QString() );
```
See note above re uuid

nyalldawson commented on src/core/qgsstoredexpressionmanager.h:
```diff
@@ -0,0 +1,158 @@
+    /**
+    * Returns the list of named expressions
+    *
+    *  \param tag               some content, maybe scope where to be shown
```
Sounds a bit unprofessional :P

nyalldawson commented on src/core/qgsvectorlayer.h:
```diff
@@ -2120,6 +2121,12 @@ class CORE_EXPORT QgsVectorLayer : public QgsMapLayer, public QgsExpressionConte
+     *
+     * \since QGIS 3.10
+     */
+    QgsStoredExpressionManager *storedFilterExpressions() { return mStoredFilterExpressions; }
```
Return reference instead of pointer?

m-kuhn replied:
It's a QObject and not copyable, hence a pointer looks just fine to me.

signedav commented on python/core/auto_generated/qgsvectorlayer.sip.in:
```diff
@@ -2279,6 +2279,12 @@ Configuration and logic to apply automatically on any edit happening on this lay
 
 
 
+    QgsStoredExpressionManager *storedFilterExpressions();
```
My idea was to name it like this, because in future maybe will follow: `storedDefaultValueExpressions(QgsStoredExpressionManager::DefaultValue)` or `storedConstraintsExpressions(QgsStoredExpressionManager::Constraints)` 

This could be done with the tags as well. Advantage of the usage of the tag could be that we can combine it better: here use constraint - and filter expressions, there use defaulvalues and constrainexpressions etc. But not sure if we need it somewhere like this. And disadvantage is that everything is in one big list. Solved as above we have seperate lists for all kinds of stored expressions.

Btw. if I do not use at all the tags in this implementation, should I remove it?

m-kuhn replied:
Is this method affected by tags? I think there will always be one manager per layer and this one manager handles the tags (or categories).

signedav replied:
Hm... so you prefer to have always one single manager containing several lists (defaultvalueexpressions, filterexpressions etc.) or one list (containing everything) but anyway categorized by the tags to identify if filter, defaultvalues etc..? 

I liked it to have several managers for several use cases, but I'm open...

m-kuhn replied:
I'd prefer to have one manager to organize the expressions. The QgsVectorLayer api is crowded enough already.

signedav committed:
- removed the modes from QgsStoredExpressionManager and do the categorization over category tags
- use QStrings instead of QUuids in QgsStoredExpression
- docs, style and beautifications

signedav committed:
sip include

nyalldawson commented on src/app/qgsattributetabledialog.cpp:
```diff
@@ -465,6 +486,47 @@ void QgsAttributeTableDialog::columnBoxInit()
+  if ( mActionHandleStoreFilterExpression->isChecked() )
+  {
+    mActionHandleStoreFilterExpression->setToolTip( tr( "Delete stored expression" ) );
+    mActionHandleStoreFilterExpression->setText( tr( "Delete Stored Expression" ) );
```
Does this show a confirmation prompt? If so, the text should end with "..."

signedav replied:
It does not show a confirmation prompt. Should it? I decided not to have one to be quick in adding/removing the stored expressions like bookmarks and because of the low importance...

nyalldawson commented on src/core/qgsstoredexpressionmanager.cpp:
```diff
@@ -0,0 +1,147 @@
+
+QString QgsStoredExpressionManager::addStoredExpression( const QString &name, const QString &expression, const QgsStoredExpression::Category &tag )
+{
+  Q_UNUSED( tag );
```
remove

signedav replied:
Uh... forgot to remove all of em.

nyalldawson commented on src/core/qgsstoredexpressionmanager.cpp:
```diff
@@ -0,0 +1,147 @@
+void QgsStoredExpressionManager::removeStoredExpression( const QString &id )
+{
+  int i = 0;
+  for ( const QgsStoredExpression &storedExpression : qgis::as_const( mStoredExpressions ) )
```
A nicer approach here would be:

    mStoredExpressions.erase( std::remove_if( mStoredExpressions.begin(), mStoredExpressions.end(), [ & ]( const QgsStoredExpression & storedExpression )
    {
      return storedExpression.id == id;
    } ), mStoredExpressions.end() );

signedav replied:
Ok, nice.

signedav replied:
I keep the original implementation, so it's same like in the other methods and easier to read.

nyalldawson commented on src/core/qgsstoredexpressionmanager.cpp:
```diff
@@ -0,0 +1,147 @@
+
+void QgsStoredExpressionManager::updateStoredExpression( const QString &id, const QString &name, const QString &expression, const QgsStoredExpression::Category &tag )
+{
+  Q_UNUSED( tag );
```
remove

nyalldawson commented on src/core/qgsstoredexpressionmanager.cpp:
```diff
@@ -0,0 +1,147 @@
+  {
+    if ( storedExpression.id == id )
+    {
+      QgsStoredExpression newStoredExpression = mStoredExpressions.at( i );
```
```suggestion
      QgsStoredExpression newStoredExpression = storedExpression;
```

nyalldawson commented on src/core/qgsstoredexpressionmanager.cpp:
```diff
@@ -0,0 +1,147 @@
+  }
+}
+
+void QgsStoredExpressionManager::addStoredExpressions( QList< QgsStoredExpression > storedExpressions )
```
```suggestion
void QgsStoredExpressionManager::addStoredExpressions( const QList< QgsStoredExpression >& storedExpressions )
```

nyalldawson commented on src/core/qgsstoredexpressionmanager.cpp:
```diff
@@ -0,0 +1,147 @@
+
+QList< QgsStoredExpression > QgsStoredExpressionManager::storedExpressions( const QgsStoredExpression::Category &tag )
+{
+  Q_UNUSED( tag );
```
I think this should be filtering for matching tags here, right?

signedav replied:
Yes, you're right.

signedav replied:
Sorry, I just see, this tag-integration has not been a very proper commit. I will update my tests for these cases as well.

nyalldawson commented on src/core/qgsstoredexpressionmanager.cpp:
```diff
@@ -0,0 +1,147 @@
+
+QgsStoredExpression QgsStoredExpressionManager::findStoredExpressionByExpression( const QString &expression, const QgsStoredExpression::Category &tag ) const
+{
+  Q_UNUSED( tag );
```
remove

nyalldawson commented on src/core/qgsstoredexpressionmanager.h:
```diff
@@ -0,0 +1,158 @@
+#ifndef SIP_RUN
+  QgsStoredExpression() = default;
+
+  QgsStoredExpression( QString name, QString expression, Category tag = Category::FilterExpression )
```
Dox

nyalldawson commented on src/core/qgsstoredexpressionmanager.h:
```diff
@@ -0,0 +1,158 @@
+  {}
+#endif
+
+  QString id;
```
Dox

nyalldawson commented on src/core/qgsstoredexpressionmanager.h:
```diff
@@ -0,0 +1,158 @@
+
+  public:
+
+    QgsStoredExpressionManager()
```
= default

nyalldawson commented on src/core/qgsstoredexpressionmanager.h:
```diff
@@ -0,0 +1,158 @@
+    *  \param name              optional name of the expression
+    *  \param expression        expression text
+    *  \param tag               category of the expression
+    *  \returns generated id as QString
```
```suggestion
    *  \returns generated id
```

nyalldawson commented on src/core/qgsstoredexpressionmanager.h:
```diff
@@ -0,0 +1,158 @@
+    void removeStoredExpression( const QString &id );
+
+    /**
+    * Updates an expression by id
```
```suggestion
    * Updates an expression by \a id.
```

signedav replied:
We tag the parameter names in the description, even when we have it described below as `\param`?

m-kuhn replied:
Yes.

nyalldawson commented on src/core/qgsstoredexpressionmanager.h:
```diff
@@ -0,0 +1,158 @@
+    */
+    QgsStoredExpression findStoredExpressionByExpression( const QString &expression, const  QgsStoredExpression::Category &tag = QgsStoredExpression::Category::FilterExpression ) const;
+
+    //! clears list of stored expressions
```
```suggestion
    //! Clears list of stored expressions
```

nyalldawson commented on src/core/qgsstoredexpressionmanager.h:
```diff
@@ -0,0 +1,158 @@
+ */
+class CORE_EXPORT QgsStoredExpressionManager : public QObject
+{
+    Q_OBJECT
```
No need for this to be a QObject - there's no signals or slots

signedav replied:
So you suggest not to use QObject and not to create a pointer from it in `QgsVectorLayer` and just return the reference, right?
That would be fine for me. We do not reassign it and we use only one instance.

@m-kuhn you defended my original approach using a pointer above. Are there any big reasons for using it versus the reference when it's not a `QObject` anymore?

m-kuhn replied:
I think we still never want a copy of it (and maybe eventually promote it to a QObject and have it send a signal when new expressions are added etc.)
So in Qt terms "it's more an identity than a value".

nyalldawson replied:
I'm fine with it being a qobject or pointer. We just need to be sure if it IS a qobject then it's parented to the layer. And also be aware that a qobject without signals/slots can throw warnings.

signedav replied:
Okay, thanks.

nyalldawson commented on src/core/qgsvectorlayer.cpp:
```diff
@@ -157,6 +158,7 @@ QgsVectorLayer::QgsVectorLayer( const QString &vectorLayerPath,
   mGeometryOptions = qgis::make_unique<QgsGeometryOptions>();
   mActions = new QgsActionManager( this );
   mConditionalStyles = new QgsConditionalLayerStyles();
+  mStoredExpressionManager = new QgsStoredExpressionManager();
```
If this IS a QObject, it should be parented to the layer and let Qt handle ownership (also correct transferal if the layer is moved between threads)

signedav committed:
proper use of category tags - use of all tag as default in filtering

signedav committed:
docs and style and set parent of mStoredExpressionManager to the layer

signedav commented:
Can we merge that @nyalldawson @m-kuhn

m-kuhn reviewed (approved):
Looks good to me, merge ahead!
